dipy/reconst/ivim.py

```diff
@@ -0,0 +1,272 @@
+    """
+    Fit the ivim params using minimize
+
+    Parameters
+    ----------
+    data : array ([X, Y, Z, ...], g)
+        Data or response variables holding the data. Note that the last
+        dimension should contain the data. It makes no copies of data.
+
+    jac : bool
+        Use the Jacobian? Default: False
+
+    Returns
+    -------
+    ivim_params: the S0, f, D_star, D value for each voxel.
+
+    """
+    flat_data = data.reshape((-1, data.shape[-1]))
+    # Flatten for the iteration over voxels:
+    bvals = gtab.bvals
+    ivim_params = np.empty((flat_data.shape[0], 4))
+
+    if routine == 'minimize':
+        result = _minimize(flat_data, bvals, x0, ivim_params,
+                           bounds, tol, jac, algorithm, gtol, ftol, eps)
+    elif routine == "leastsq":
+        result = _leastsq(flat_data, bvals, x0, ivim_params)
+    ivim_params.shape = data.shape[:-1] + (4,)
+    return ivim_params
+
+
+def _minimize(flat_data, bvals, x0, ivim_params,
+              bounds, tol, jac, algorithm, gtol, ftol, eps):
+    """Use minimize for finding ivim_params"""
+    sum_sq = lambda params, bvals, signal: np.sum(
+        _ivim_error(params, bvals, signal))
+
+    result = []
+    for vox in range(flat_data.shape[0]):
+        res = minimize(sum_sq,
+                       x0[vox],
+                       args=(bvals, flat_data[vox]), bounds=bounds,
+                       tol=tol, method=algorithm, jac=jac,
+                       options={'gtol': gtol, 'ftol': ftol, 'eps': eps})
+        ivim_params[vox, :4] = res.x
+        result += res
+    return result
+
+
+def _leastsq(flat_data, bvals, x0, ivim_params):
+    """Use minimize for finding ivim_params"""
+    result = []
+    for vox in range(flat_data.shape[0]):
+        res = leastsq(_ivim_error,
+                      x0[vox],
+                      args=(bvals, flat_data[vox]))
+        ivim_params[vox, :4] = res[0]
+        result += res
+    return result
+
+
+def two_stage(data, gtab, x0,
+              split_b, jac, bounds, tol,
+              routine, algorithm, gtol, ftol, eps):
+    """
+    Fit the ivim params using a two stage fit
+
+    Parameters
+    ----------
+    data : array ([X, Y, Z, ...], g)
+        Data or response variables holding the data. Note that the last
+        dimension should contain the data. It makes no copies of data.
+
+    jac : bool
+        Use the Jacobian? Default: False
+
+    Returns
+    -------
+    ivim_params: the S0, f, D_star, D value for each voxel.
+
+    """
+    flat_data = data.reshape((-1, data.shape[-1]))
+    # Flatten for the iteration over voxels:
+    bvals_ge_split = gtab.bvals[gtab.bvals > split_b]
+    bvecs_ge_split = gtab.bvecs[gtab.bvals > split_b]
+    gtab_ge_split = gradient_table(bvals_ge_split, bvecs_ge_split.T)
+
+    tensor_model = TensorModel(gtab_ge_split)
+    tenfit = tensor_model.fit(data[..., gtab.bvals > split_b])
+
+    D_guess = mean_diffusivity(tenfit.evals)
+
+    x0[:, 3] = D_guess
```

Comment:
Here's the real equation: `f_guess = 1 - S0' / S0` where `S0'` is the diffusion b>cutoff 0 intercept, and `S0` is the perfusion b<=cutoff 0 intercept. So `S0'` is smaller than `S0` unless the data are noisy or otherwise strange.